<!--
%\VignetteIndexEntry{02.2 Other Sequence Analysis Work Flows}
%\VignettePackage{LearnBioconductor}
%\VignetteEngine{knitr::knitr}
-->

```{r setup, echo=FALSE}
library(LearnBioconductor)
stopifnot(BiocInstaller::biocVersion() == "3.0")
```

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
knitr::opts_chunk$set(cache=TRUE, tidy=FALSE)
```

# Common Sequence Analysis Work Flows

Martin Morgan, Sonali Arora<br/>
October 28, 2014

## ChIP-seq

See my recent
[slides](http://bioconductor.org/help/course-materials/2014/CSAMA2014/4_Thursday/lectures/ChIPSeq_slides.pdf)
outlining ChIP-seq and relevant _Bioconductor_ software.

## Variants

See Michael Lawrence's variant calling with
[VariantTools](http://bioconductor.org/help/course-materials/2014/BioC2014/Lawrence_Tutorial.pdf).
and Val Obenchain's manipulation and annotation of called variants with
[VariantAnnotation](http://bioconductor.org/help/workflows/variants/).

## Copy Number

See the [Copy Number Workflow](./B02.2.3_CopyNumber.html) document.

## Epigenomics

See the short
[introduction](http://bioconductor.org/help/course-materials/2014/Epigenomics/MethylationArrays.html)
and
[lab](http://bioconductor.org/help/course-materials/2014/Epigenomics/MethylationArrays-lab.html)
centered around Illumina 450k methylation arrays and the `r Biocpkg("minfi")` package.
